File: sdks/python/harness_common/service.py
"""A Component's runtime presence in one Environment."""

from __future__ import annotations

from dataclasses import dataclass, field

from harness_common.component import Component
from harness_common.environment import Environment
from harness_common.workload import Workload


@dataclass(frozen=True, slots=True)
class Service:
    """One Component's logical runtime presence in an Environment.

    workloads declares zero or more platform carriers; it is neither a live Pod
    inventory nor proof of existence. The mapping may change without changing
    the logical Service identity. Consumers choose sampling/aggregation policy.
    """

    name: str
    component: Component
    environment: Environment

    workloads: tuple[Workload, ...] = field(default=(), kw_only=True, compare=False)

File: sdks/typescript/common/src/service.ts
import type { Component } from "./component.js";
import type { Environment } from "./environment.js";
import type { Workload } from "./workload.js";

/**
 * One Component's logical runtime presence in an Environment.
 * @spec Identity is scoped by name, Component and Environment, not platform resource names.
 * @rule Workload mappings may change without changing Service identity; declarations prove no live instances.
 * Consumers extend this model with their own capabilities and execution policy.
 */
export interface Service<E extends Environment = Environment> {
  readonly name: string;
  readonly component: Component;
  readonly environment: E;
  /** Zero or more runtime carriers; an empty array declares no workloads. */
  readonly workloads: readonly Workload[];
}
